### ISSUE TITLE
OnHeapStore.setAccessTimeAndExpiryThenReturnMapping() calls valueHolder.value()

### ISSUE THREAD

**Issue by lorban** [ehcache/ehcache3#790]
This is required per the `Expiry` contract to pass the value to the expiry instance.

The nasty drawback is that this makes deserialization happen twice: once to pass to the expiry and once in `Ehcache.get()` to return to the user call.

Deserializing twice has obvious disastrous performance consequence, especially since our default `Expiry` implementation has no use for the deserialized value.

One thing that could be done is change the `Expiry` contract to pass `Holder<V>` instead of `V` to its methods so that only the expiries making use of the value would pay the extra deserialization price.

**Comment by ljacomet** [comment]
So this is not as easy as it sounds.

* To be really efficient, not only the expiry contract has to be adapted, the events one too.
 * to be fair the event one could be an internal adaptation only, not exposed to `CacheEventListener`
* Doing this change actually adds overhead on create events which usually have the value in _clear_ and it would now have to be wrapped
* Generics are a mess :cry:

**Comment by ljacomet** [comment]
I will soon push a PR for these changes. I really do not like the results and what we have to do to enable this.

So I will be looking for conversation on this topic and most likely wait for all others perf issues to be resolved and see how much we are talking about exactly here.

**Comment by ljacomet** [comment]
Awaiting more performance results after #777, #787 and #799 have been applied.

**Comment by ljacomet** [comment]
Realised that the changes in the initial version of the pull request were badly broken for offheap where value holders are now lazy. So the change is that in all methods where the value needs to be returned, it has to be forced to deserialize or detached because expiry computation no longer triggers it. And remember that off heap will only deserialize a value once since the reference has a life limited to the invoking thread. So while there will be differences, I am afraid they will be minimal.

On heap the discussion is a bit different because values are deserialized at each invocation indeed. But that is limited to use cases where the heap tier is configured with a `SerializingCopier` which is clearly NOT the recommended way if you care for maximum performance.

So all this really makes me doubt this PR even more.

**Comment by lorban** [comment]
Except that JSR107 caches enable store-by-value by default, I do agree with the above.

We are close to our last opportunity to change the API to make this possible, even if we decide to always deserialize for now and postpone the optimization.

Food for thoughts...

**Comment by chrisdennis** [comment]
I like these changes, they seem to make the API more honest. Retrieving the values is something being done that may otherwise not be done... so it has a cost. Associating a cost with an action is a nice way of communicating to the user that what they are doing might have a performance impact.

**Comment by ljacomet** [comment]
Given the comments from @lorban and @chrisdennis, let's go with this.
